### PR TITLE
impl(generator): format discovery JSON to protobuf message

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #10980

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119)

### CODE REVIEW

File: generator/internal/discovery_type_vertex.cc [hunk 1]
```diff
@@ -151,6 +152,86 @@ DiscoveryTypeVertex::DetermineTypeAndSynthesis(nlohmann::json const& v,
...
+  }
+  int field_number = 1;
+  std::string indent(indent_level * 2, ' ');
+  std::vector<std::string> text;
+  auto const& properties = json.find("properties");
+  for (auto p = properties->begin(); p != properties->end(); ++p) {
+    auto const& v = p.value();
+    std::string field_name = v.value("id", p.key());
+    auto type_and_synthesize = DetermineTypeAndSynthesis(v, field_name);
+    if (!type_and_synthesize) return type_and_synthesize.status();
```
Comment (dbolduc) [inline generator/internal/discovery_type_vertex.cc:185]: std::move

File: generator/internal/discovery_type_vertex.cc [hunk 2]
```diff
@@ -151,6 +152,86 @@ DiscoveryTypeVertex::DetermineTypeAndSynthesis(nlohmann::json const& v,
...
+
+    std::string type = type_and_synthesize->name;
+    if (type_and_synthesize->properties != nullptr) {
+      auto result =
+          FormatMessage(type, *type_and_synthesize->properties, indent_level);
+      if (!result) return result.status();
+      text.push_back(*result);
```
Comment (dbolduc) [inline generator/internal/discovery_type_vertex.cc:192]: std::move(result)

File: generator/internal/discovery_type_vertex.cc [hunk 3]
```diff
@@ -151,6 +152,86 @@ DiscoveryTypeVertex::DetermineTypeAndSynthesis(nlohmann::json const& v,
...
+  std::vector<std::string> text;
+  auto const& properties = json.find("properties");
+  for (auto p = properties->begin(); p != properties->end(); ++p) {
+    auto const& v = p.value();
+    std::string field_name = v.value("id", p.key());
+    auto type_and_synthesize = DetermineTypeAndSynthesis(v, field_name);
+    if (!type_and_synthesize) return type_and_synthesize.status();
+
+    std::string type = type_and_synthesize->name;
+    if (type_and_synthesize->properties != nullptr) {
```
Comment (dbolduc) [inline generator/internal/discovery_type_vertex.cc:188]: opt: `s/ != nullptr//`

**Review by scotthart — COMMENTED** [review]
Applied said fix throughout the file.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-:-NRdEnjY62fU1SU3yipQ:brotq7a)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @dbolduc)

___
*[`generator/internal/discovery_type_vertex.cc` line 165 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-165:-NRdBnonE3u7zwUlOLxM:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L165)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit:: `std::move(lines).status()`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 185 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-185:-NRdBteR6C8e3aD0IahF:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L185)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

std::move
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 188 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-188:-NRdCW7U6DgH_xd0wRFQ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L188)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

opt: `s/ != nullptr//`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 192 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-192:-NRdCgp99kWqjtNsRi6n:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L192)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

std::move(result)
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 222 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-222:-NRdCoPe1QQIXA3OBx83:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L222)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

std::move()
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 272 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-272:-NRdCpSH9MUWYdkmL7G1:b74zx7t) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L272)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

do we want a test for the case where `json_.contains("description") == false` ?
</blockquote></details>

Added test case to make sure we test for both branches.
___
*[`generator/internal/discovery_type_vertex.cc` line 279 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-279:-NRdDiUG8BxeDqc6s-K1:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L279)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: `std::move(message).status()`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 280 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFv-r2-280:-NRdDjIJBv4pmNY17tmo:b-c2zwn8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex.cc#L280)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: `*std::move(message)` ?
</blockquote></details>

StrAppend always has to copy into the 1st argument. It doesn't benefit from `std::move()`'ing any of the subsequent arguments.
___
*[`generator/internal/discovery_type_vertex_test.cc` line 321 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11119#-NRZ9HT1qe9GLp7aSJFx-r2-321:-NRdEDNeB5z4Y6tTkbof:bsv8dpb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/68324d357f71e5cd18f90c1fca2a28dfe88d6476/generator/internal/discovery_type_vertex_test.cc#L321)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: `EXPECT_EQ(*result, kExpectedProto)` ?

here and elsewhere
</blockquote></details>

Use of EXPECT_THAT(, Eq) is locally consistent and intentional as failures in an EXPECT_THAT has slightly better formatted diagnostics.

**Pull request merged**